src/com/palmergames/bukkit/towny/utils/SpawnUtil.java

```diff
@@ -0,0 +1,395 @@
+
+		// Disallow jailed players from teleporting.
+		if (resident.isJailed())
+			throw new TownyException(TownySettings.getLangString("msg_cannot_spawn_while_jailed"));
+
+		Town town = null;
+		Nation nation = null;
+		Location spawnLoc = null;
+		TownSpawnLevel townSpawnPermission = null;
+		NationSpawnLevel nationSpawnPermission = null;
+		boolean isTownyAdmin = townyUniverse.getPermissionSource().has(player, spawnType.getNode());
+
+		// Figure out which Town/NationSpawnLevel this is.
+		// Resolve where the spawnLoc will be.
+		switch (spawnType) {
+		case RESIDENT:
+			if (resident.hasTown())
+				town = resident.getTown();
+			if (TownySettings.getBedUse() && player.getBedSpawnLocation() != null)
+				spawnLoc = player.getBedSpawnLocation();
+			else if (town != null)
+				spawnLoc = town.getSpawn();
+			else
+				spawnLoc = plugin.getCache(player).getLastLocation().getWorld().getSpawnLocation();
+
+			if (isTownyAdmin) {
+				townSpawnPermission = TownSpawnLevel.ADMIN;
+			} else {
+				townSpawnPermission = TownSpawnLevel.TOWN_RESIDENT;
+			}
+			break;
+
+		case TOWN:
+			town = (Town) townyObject;
+			if (outpost) {
+				if (!town.hasOutpostSpawn())
+					throw new TownyException(TownySettings.getLangString("msg_err_outpost_spawn"));
+
+				Integer index = null;
+				try {
+					if (!split[split.length - 1].contains("name:")) {
+						index = Integer.parseInt(split[split.length - 1]);
+					} else { // So now it say's name:123
+						split[split.length - 1] = split[split.length - 1].replace("name:", "").replace("_", " ");
+						for (Location loc : town.getAllOutpostSpawns()) {
+							TownBlock tboutpost = TownyAPI.getInstance().getTownBlock(loc);
+							if (tboutpost != null) {
+								String name = tboutpost.getName();
+								if (name.startsWith(split[split.length - 1])) {
+									index = 1 + town.getAllOutpostSpawns().indexOf(loc);
+								}
+							}
+						}
+						if (index == null) { // If it persists to be null, so it's not been given a value, set it to the
+												// fallback (1).
+							index = 1;
+						}
+					}
+				} catch (NumberFormatException e) {
+					// invalid entry so assume the first outpost, also note: We DO NOT HAVE a number
+					// now, which means: if you type abc, you get brought to that outpost.
+					// Let's consider the fact however: an outpost name begins with "123" and there
+					// are 123 Outposts. Then we put the prefix name:123 and that solves that.
+					index = 1;
+					// Trying to get Outpost names.
+					split[split.length - 1] = split[split.length - 1].replace("_", " ");
+					for (Location loc : town.getAllOutpostSpawns()) {
+						TownBlock tboutpost = TownyAPI.getInstance().getTownBlock(loc);
+						if (tboutpost != null) {
+							String name = tboutpost.getName();
+							if (name.startsWith(split[split.length - 1])) {
+								index = 1 + town.getAllOutpostSpawns().indexOf(loc);
+							}
+						}
+					}

```

Comment:
This can be cleaned up. 
First, just parse for an integer and catch the numberformat exception. That way we don't have to preemptively duplicate a check for a name. 
Then since we are going to check for name anyway, just replace `name:` from the string in case it exists.

Also instead of calling `town.getAllOutSpawns().indexOf()` which will basically loop through all of the locs again, use a for-i loop. 

Thus the new code would look something like:
```java
int index = 1;
String outpostName = split[split.length -1];

try {
	index = Integer.parseInt(outpostName);
} catch (NumberFormatException nfe) {
	outpostName = outpostName.replace("name:", "").replace("_", " ");

	for (int outpostLocIndex = 0; outpostLocIndex < town.getAllOutpostSpawns().size(); outpostLocIndex++) {
		Location outpostLoc = town.getAllOutpostSpawns().get(outpostLocIndex);
		TownBlock tbOutpost = TownyAPI.getInstance().getTownBlock(outpostLoc);

		if(tbOutpost != null && tbOutpost.getName.startsWith(outpostName) {
			index += outpostLocIndex; // Remember index is already set to 1
			break;
		}
	}
}